### ISSUE TITLE
Perl ZooKeeper locks up during heavy load

### ISSUE THREAD

**Issue by kvanhorn2015** [mark-5/p5-net-zookeeper#5]
We are using Perl Net::ZooKeeper (0.38) and Net::ZooKeeper::Lock (0.03) libraries. Deadlock appears to occur at the end during lock cleanup activity. Here is a stack dump (sensitive names changed):

Thread 2 (Thread 0x2ac6fbfa3940 (LWP 13292)):
#0 0x00002ac6f5aed654 in __lll_lock_wait () from /lib64/libpthread.so.0
#1 0x00002ac6f5aeb47b in pthread_cond_signal@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#2 0x00002ac6f835539c in _zk_watcher (handle=<value optimized out>, type=2, state=3, path=<value optimized out>, context=0x33f3ce0) at ZooKeeper.xs:179
#3 0x00002ac6f856d942 in do_foreach_watcher (zh=0x33e4fb0, type=2, state=3, path=0x33f3f50 "/lock/cmts/cisco_device1.net-0001851215", list=0x33ed290)
at /home/myhome/rpm/BUILD/zookeeper-3.4.6/src/c/src/zk_hashtable.c:279
#4 deliverWatchers (zh=0x33e4fb0, type=2, state=3, path=0x33f3f50 "/lock/cmts/cisco_device1.net-0001851215", list=0x33ed290)
at /home/myhome/rpm/BUILD/zookeeper-3.4.6/src/c/src/zk_hashtable.c:321
#5 0x00002ac6f8564966 in process_completions (zh=0x33e4fb0) at /home/myhome/rpm/BUILD/zookeeper-3.4.6/src/c/src/zookeeper.c:2114
#6 0x00002ac6f856e101 in do_completion (v=<value optimized out>) at /home/myhome/rpm/BUILD/zookeeper-3.4.6/src/c/src/mt_adaptor.c:466
#7 0x00002ac6f5ae683d in start_thread (arg=<value optimized out>) at pthread_create.c:301
#8 0x00002ac6f5dd1fcd in clone () from /lib64/libc.so.6

Thread 1 (Thread 0x2ac6f6056af0 (LWP 12972)):
#0 0x00002ac6f5ae7c65 in pthread_join (threadid=47034119371072, thread_return=0x0) at pthread_join.c:89
#1 0x00002ac6f856e7de in adaptor_finish (zh=0x33e4fb0) at /home/myhome/rpm/BUILD/zookeeper-3.4.6/src/c/src/mt_adaptor.c:293
#2 0x00002ac6f8566cdc in zookeeper_close (zh=0x33e4fb0) at /home/myhome/rpm/BUILD/zookeeper-3.4.6/src/c/src/zookeeper.c:2536
#3 0x00002ac6f8357222 in XS_Net__ZooKeeper_DESTROY (my_perl=0x20df010, cv=<value optimized out>) at ZooKeeper.xs:885
#4 0x00002ac6f4b38af6 in Perl_pp_entersub () from /usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi/CORE/libperl.so
#5 0x00002ac6f4adb8d7 in ?? () from /usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi/CORE/libperl.so
#6 0x00002ac6f4adf720 in Perl_call_sv () from /usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi/CORE/libperl.so
#7 0x00002ac6f4b3d3c6 in Perl_sv_clear () from /usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi/CORE/libperl.so
#8 0x00002ac6f4b3db70 in Perl_sv_free () from /usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi/CORE/libperl.so
#9 0x00002ac6f4b6025c in Perl_free_tmps () from /usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi/CORE/libperl.so
#10 0x00002ac6f4adf78a in Perl_call_sv () from /usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi/CORE/libperl.so
#11 0x00002ac6f4b3d3c6 in Perl_sv_clear () from /usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi/CORE/libperl.so
#12 0x00002ac6f4b3db70 in Perl_sv_free () from /usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi/CORE/libperl.so
#13 0x00002ac6f4b3b0e5 in ?? () from /usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi/CORE/libperl.so
#14 0x00002ac6f4b3b141 in Perl_sv_clean_objs () from /usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi/CORE/libperl.so
#15 0x00002ac6f4ae185e in perl_destruct () from /usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi/CORE/libperl.so
#16 0x0000000000401773 in main ()

There are about 4 out of 10,000 processes that end up in deadlock, and according to our web searches, the only reason pthread_cond_signal would lock is due to it waiting on an already destroyed condition.

As a cross-reference, a Jira ticket on apache.org has been created [here] (https://issues.apache.org/jira/browse/ZOOKEEPER-2200?jql=project%20%3D%20ZOOKEEPER%20AND%20text%20~%20perl).

**Comment by mark-5** [comment]
I looked a bit into this, and I did see one way to deadlock in Net::ZooKeeper. Can you tell me if this is likely the issue you are experiencing? I'll explain more below...

The [ZooKeeper watcher docs](http://zookeeper.apache.org/doc/r3.4.6/zookeeperProgrammers.html#sc_WatchRememberThese) explain that all watches are triggered on disconnect.
> When you disconnect from a server (for example, when the server fails), you will not get any watches until the connection is reestablished. For this reason session events are sent to all outstanding watch handlers. Use session events to go into a safe mode: you will not be receiving events while disconnected, so your process should act conservatively in that mode.

When watches are triggered for these session events, ZooKeeper still considers them live, and watches can still be triggered by the watcher event or future session events. So one scenario that can cause deadlocks([here](https://gist.github.com/mark-5/b2c66897ee27b1bd017d) is an example gist) can be reproduced as follows:

1. a watch is registered
2. a disconnect triggers the watch for a session event
3. a new watch is registered(this causes Net::ZooKeeper to clean up old watches)
4. the first watch is triggered(either by the watch event or a session event)

What is happening is that the session event is triggering a watch, and Net::ZooKeeper will prematurely destroy the watch and the condition variable. The watch is then triggered later again, and tries to use the now destroyed condition variable.

Does this seem plausible for your setup? Do you know if your clients are getting temporarily disconnected from the server?

**Comment by kvanhorn2015** [comment]
Thanks, we were able to verify that all cases of deadlock in the past 2 days happened on a re-established  session. Example from the log:

2015-06-16 15:07:55,188 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:ZooKeeperServer@861] - Client attempting to renew session 0x14cc4bdc35c8f93 at /127.0.0.1:30539

Also, we were able to replicate the issue with these steps:
1. Acquire a lock (Net::ZooKeepr::Lock) in process1.
2. Attempt to acquire a lock in process2.
3. Using iptables break network connection to the ZooKeeper server for process2
4. Process1 completes and releases lock
5. Process2 acquires locks, finishes and attempts to remove lock.
6. Process2 hangs indefinitely until killed.
